### DISCUSSION TITLE
Batteriewerte vom Zendure-Solarflow via MQTT aus Zendure-Cloud auslesen

### DISCUSSION THREAD

**Discussion by MaXxGrr** [hoylabs/OpenDTU-OnBattery#898] [Q&A]
Hallo @kutzera!

Wie ich sehe hast Du den Ladezustand vom Zendure-Akku über MQTT von der Zendure-Cloud in OpenDTUoB reinbekommen: https://github.com/helgeerbe/OpenDTU-OnBattery/discussions/712#discussioncomment-8737376 

Ich hab mit dieser Anleitung hier https://github.com/z-master42/solarflow/blob/main/solarflow.md die Solarflow-Cloud per MQTT angezapft und in HomeAssistant eingebunden. Allerdings hab ich bei den MQTT-Topics und JSON-Formaten keinen Überblick und auch keine weiterführende Idee außer sinnloses testen. 

Bei den Batterie-Einstellungen in OpenDTU-oB habe ich versuchsweise "Batteriewerte aus MQTT-Broker" ausgewählt und als Topic den richtigen rausgesucht und eingetragen (PgB307Lf/p7fG05LE/state). Der Topic kommt auch in OpenDTU-oB an und wird ihm Log angezeigt, allerdings klappt das parsen des Wertes nicht: 

```
11:44:52.069 > Received MQTT message on topic: PgB307Lf/p7fG05LE/state
11:44:52.214 > MqttBattery: Cannot parse payload '{"electricLevel":42,"packData":[{"sn":"XO4OHGHOLK14547"}],"sn":"XO1HKEXLH102939"}' in topic 'PgB307Lf/p7fG05LE/state' as float
```

Wenn ich beim Topic `PgB307Lf/p7fG05LE/state "{{ value_json.electricLevel | int }}"` eintrage bekomme ich überhaupt nix zu MQTT, Battery oder so angezeigt. Wenn ich irgendwas ähnliches ohne Leerzeichen an den Topic anhänge, (hab es auch schon ohne Anführungszeichen, mit Hochkommata, etc. probiert) bekomme ich im Log lediglich folgenden Fehler angezeigt: 
```
11:59:11.615 > Connecting to MQTT...
11:59:11.798 > Connected to MQTT.
11:59:12.023 > Disconnected from MQTT.
11:59:12.229 > Disconnect reason:TCP_DISCONNECTED
```
oder gleich gar keine relevanten Log-Eintrag... Insofern fehlt mir der Ausdruck, der den in der Payload ja vorhandenen Wert `electricLevel` korrekt parst. Die Serial-Nr. müßte bei mir gar nicht ausgewertet werden, da ich nur einen Zendure-Akku hab. 

Vielen Dank schonmal im Voraus, MaXx

**Comment by kutzera** [comment]
Hi!Versuch mal bitte eine Mqtt.yaml mit den Werten aus dem z-Master42 Beitrag auf Github zu erstellen. Dadurch werden die Zendure  Entitäten angelegt. Danach habe ich einen Helfer mit den Akkuwerten angelegt, da ich zwei Speicher kombiniere. Dieser wird über den MQTT-Broker und eine Automation gepublished. Freundliche GrüßeSebastianAm 20.04.2024 um 15:16 schrieb MaXx ***@***.***>:﻿
Hallo @kutzera!
Wie ich sehe hast Du den Ladezustand vom Zendure-Akku über MQTT von der Zendure-Cloud in OpenDTUoB reinbekommen: #712 (reply in thread)
Ich hab mit dieser Anleitung hier https://github.com/z-master42/solarflow/blob/main/solarflow.md die Solarflow-Cloud per MQTT angezapft und in HomeAssistant eingebunden. Allerdings hab ich bei den MQTT-Topics und JSON-Formaten keinen Überblick und auch keine weiterführende Idee außer sinnloses testen.
Bei den Batterie-Einstellungen in OpenDTU-oB habe ich versuchsweise "Batteriewerte aus MQTT-Broker" ausgewählt und als Topic den richtigen rausgesucht und eingetragen (PgB307Lf/p7fG05LE/state). Der Topic kommt auch in OpenDTU-oB an und wird ihm Log angezeigt, allerdings klappt das parsen des Wertes nicht:
11:44:52.069 > Received MQTT message on topic: PgB307Lf/p7fG05LE/state
11:44:52.214 > MqttBattery: Cannot parse payload '{"electricLevel":42,"packData":[{"sn":"XO4OHGHOLK14547"}],"sn":"XO1HKEXLH102939"}' in topic 'PgB307Lf/p7fG05LE/state' as float

Wenn ich beim Topic PgB307Lf/p7fG05LE/state "{{ value_json.electricLevel | int }}" eintrage bekomme ich überhaupt nix zu MQTT, Battery oder so angezeigt. Wenn ich irgendwas ähnliches ohne Leerzeichen an den Topic anhänge, (hab es auch schon ohne Anführungszeichen, mit Hochkommata, etc. probiert) bekomme ich im Log lediglich folgenden Fehler angezeigt:
11:59:11.615 > Connecting to MQTT...
11:59:11.798 > Connected to MQTT.
11:59:12.023 > Disconnected from MQTT.
11:59:12.229 > Disconnect reason:TCP_DISCONNECTED

oder gleich gar keine relevanten Log-Eintrag... Insofern fehlt mir der Ausdruck, der den in der Payload ja vorhandenen Wert electricLevel korrekt parst. Die Serial-Nr. müßte bei mir gar nicht ausgewertet werden, da ich nur einen Zendure-Akku hab.
Vielen Dank schonmal im Voraus, MaXx

**Reply by MaXxGrr** [reply]
Danke für Deine Antwort - ich hatte gehofft, daß Du das ohne Zwischeninstanz hinbekommen hast. Weil eigentlich ist das MQTT-Topic ja im Netzwerk vorhanden, und eigentlich versuche ich überflüssige Zwischenschritte weg zu lassen. Bei den Einstellungen zu den Stromzählern gibt's ja auch MQTT und anscheinend ist auch eine FirebaseJson-Bibliothek an Bord, die JSON parsen können sollte. 

Bei der Bibliothek hab ich aber nicht ganz verstanden, wie die das JSON der originalen MQTT-Nachricht parsen kann. Bzw. wie ich das eingeben muß... 

Vielleicht hat ja jemand anderes hier mit dem Parsen von JSON in MQTT-Nachrichten schon Erfolg gehabt... 

Grüazi, MaXx

**Comment by JochenKlenk** [comment]
Ich hab das gleiche Problem. Daten kommen vom mqtt in Homeassitant.

11:40:17.701 > Received MQTT message on topic: noah2mqtt/0PVP20ZR15JT00BT
11:40:17.757 > [MqttBattery] Topic 'noah2mqtt/0PVP20ZR15JT00BT': cannot parse payload '{"output_w":38,"solar_w":56,"soc...' as float
11:40:17.805 > [MqttBattery] Topic 'noah2mqtt/0PVP20ZR15JT00BT': cannot parse payload '{"output_w":38,"solar_w":56,"soc...' as float
11:40:17.849 > Interrupt received

Jemand eine Idee?

Danke schon mal

**Reply by spcqike** [reply]
was hast du denn konfiguriert? Ich nehme an, die Batterie ist als MQTT vom Broker eingestellt.

dein SoC Topic ist .... "noah2mqtt/0PVP20ZR15JT00BT" und dein JSON-Pfad "soc"?

**Reply by JochenKlenk** [reply]
der ganze Output kommt als ein Payload und wie im Auszug ganz am Ende zu sehen ist das Topix dann "soc" das gebraucht würde.

Hab auch schon das hier versucht aber dann sehe ich gar nichts mehr in den Logs vom MQTT connect.
noah2mqtt/0PVP20ZR15JT00BT "{{ value_json.soc | int }}"

Das ganze kommt so vom Homeassitant

Payload: {"output_w":0,"solar_w":43,"soc":19,"charge_w":25,"discharge_w":0,"battery_num":1,"generation_total_kwh":148.7,"generation_today_kwh":0.3,"work_mode":"battery_first"}

Danke für jede Hilfe.

**Reply by spcqike** [reply]
> dein SoC Topic ist .... "noah2mqtt/0PVP20ZR15JT00BT" und dein JSON-Pfad "soc"?

hast du das schon probiert?
weil, genau das sollte es sein. wenn es denn richtig geht und kein Bug gibt. vermute ich

**Reply by JochenKlenk** [reply]
Habe es mehrmals getestet nun noch einmal und den ganzen ESP stromlos und neugestartet und es geht.

Danke für den Support
